### PR TITLE
[fix] composite compliance: margin_ranking_loss, hinge_embedding_loss

### PR DESCRIPTION
**Pull request by kshitij12345**
Ref: #69991 

Cause of failure is similar to the one discussed for fixing forward_ad of `nn.functional.linear`: https://github.com/pytorch/pytorch/pull/77950#discussion_r878328822

### CODE REVIEW

File: aten/src/ATen/native/Loss.cpp [hunk 1]
```diff
@@ -139,7 +139,14 @@ Tensor cosine_embedding_loss(const Tensor& input1, const Tensor& input2, const T
...
   auto zeros = at::zeros_like(self, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
-  auto margin_clamp = (margin - self).clamp_min_(0);
+  auto margin_diff = (margin - self);
+  // For Composite Compliance,
+  // In Forward AD, if `margin_diff` is a CCT but its tangent isn't,
+  // using inplace clamp_min doesn't work.
+  auto margin_clamp = (margin_diff._fw_grad(/*level*/ 0).defined() &&
+                       isTensorSubclassLike(margin_diff))
+      ? margin_diff.clamp_min(0)
+      : margin_diff.clamp_min_(0);
```
Comment (zou3519) [inline aten/src/ATen/native/Loss.cpp:149]: I'm a little confused. Let's say margin_diff is a CCT but its tangent isn't. Then in the original code,
```
auto margin_clamp = margin_diff.clamp_min_(0);
```
What I would expect the output to be is:
- margin_clamp is a CCT
- margin_clamp's tangent is a regular Tensor

What is the problem here? It seems OK to me that the tangent isn't a CCT, but I'm probably missing something

Comment (kshitij12345) [reply aten/src/ATen/native/Loss.cpp:149]: https://github.com/pytorch/pytorch/blob/bb3e1f30a8a4d04c501ce01d15df8a84b5feb91c/tools/autograd/derivatives.yaml#L382-L384

Assuming `self` is CCT and looking at the formula for `clamp_min`,
we can see that it uses `self` which leads to the resultant `tangent` to be `CCT` which we try to copy into a regular tensor.

Comment (zou3519) [reply aten/src/ATen/native/Loss.cpp:149]: Where is the "copy_ into tangent" happening?

**Review by zou3519 — APPROVED** [review]
LGTM, one minor nit to add a little more detail to the comment
